url587829ef81_champdavis_KhovanovHomology_tgz-pJ8-KhovanovMapJ8/src/KhovanovMap.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.commons.lang3.StringUtils;
+import java.util.*;
+
+public class KhovanovMap{
+    
+    private Knot theKnot;  //root knot
+    private Knot[] resolvedKnots;
+    private KhovanovRow[] rows; 
+    private Stack<Map> maps;
+
+    
+    public KhovanovMap(Knot kn){
+        this.theKnot = kn;
+        this.rows = new KhovanovRow[this.theKnot.numCrossings()+1];
+        
+        int n = this.rows.length-1; //COMBINATION FINDER since length of each row is based on combinations....it's nCk in this case
+        int nCk = 1;
+        for (int k = 0; k <= n; k++) {
+            this.rows[k] = new KhovanovRow(nCk);
+            nCk = nCk * (n-k) / (k+1);
+        }
+            
+        //this.resolvedKnots = new Knot[(int) Math.pow(2,this.theKnot.numCrossings())];
+        
+        fillResolvedKnots(this.theKnot);
+        setMaps(); 
+            
+    }
+    
+    public void fillResolvedKnots(Knot k){
+        if(k.getLeftKnot()==null){
+            int numOnes = StringUtils.countMatches(k.getBinCode(), "1");
+            this.rows[numOnes].add(new KhovanovMapNode(k.getBinCode(),k.getCircles()));
+            //this.resolvedKnots[Integer.parseInt(k.getBinCode(), 2)] = k; //puts the knot in its right place in the array based on its binary code
+        }
+            
+
+        else{
+            fillResolvedKnots(k.getLeftKnot());
+            fillResolvedKnots(k.getRightKnot());
+        }
+    }
+    
+    public void setMaps(){       
+        KhovanovMapNode current;
+        MergeMap merge;
+        SplitMap split;        
+        
+        for(int i=0; i<this.rows.length; i++){
+            for(int k=0; k<this.rows[i].size(); k++){
+                current = this.rows[i].nthNode(k);
+                String[] outMaps = current.outMaps();
+                for(String s:outMaps){
+                    KhovanovMapNode to = findNode(s);
+
+                    ArrayList<SeifertCircle> fromCircles = new ArrayList<SeifertCircle>();
+                    ArrayList<SeifertCircle> toCircles = new ArrayList<SeifertCircle>();
+                    ArrayList<SeifertCircle> sameCircles = new ArrayList<SeifertCircle>();
+                        
+                    SeifertCircle theCircle; // to be passed into the map
+                    if(current.numCircles() < to.numCircles()){ //split map
+                        theCircle = new SeifertCircle("");
+                        fromCircles = current.getCircles();
+                        toCircles = to.findDifferentCircles(current);
+                        for(SeifertCircle sei:fromCircles){
+                            sameCircles = to.findCircleMatch(sei);
+                            if(sameCircles.size()==0){
+                                theCircle = sei;
+                                break;
+                            }
+                        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 65:71, 89:95.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.